editor/surface_upgrade_dialog.cpp

```diff
@@ -0,0 +1,91 @@
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "surface_upgrade_dialog.h"
+
+#include "editor/editor_file_system.h"
+#include "editor/editor_node.h"
+#include "servers/rendering_server.h"
+
+void SurfaceUpgradeDialog::_add_files(EditorFileSystemDirectory *p_dir, HashSet<String> &p_paths, PackedStringArray &p_files) {
+	for (int i = 0; i < p_dir->get_subdir_count(); i++) {
+		_add_files(p_dir->get_subdir(i), p_paths, p_files);
+	}
+
+	for (int i = 0; i < p_dir->get_file_count(); i++) {
+		if (p_dir->get_file_type(i) == "Mesh" ||
+				p_dir->get_file_type(i) == "ArrayMesh" ||
+				p_dir->get_file_type(i) == "PackedScene") {
+			if (FileAccess::exists(p_dir->get_file_path(i) + ".import")) {
+				p_files.push_back(p_dir->get_file_path(i));
+			} else {
+				p_paths.insert(p_dir->get_file_path(i));
+			}
+		}
+	}
+}
+
+void SurfaceUpgradeDialog::upgrade_all_meshes() {
+	// Update all meshes here.
+	HashSet<String> paths;
+
+	PackedStringArray files_to_import = PackedStringArray();
+
+	_add_files(EditorFileSystem::get_singleton()->get_filesystem(), paths, files_to_import);
+
+	EditorProgress ep("Re-saving all scenes and meshes", TTR("Upgrading All Meshes in Project"), paths.size());
+
+	ep.step(TTR("Re-importing meshes"), 0);
+	EditorFileSystem::get_singleton()->reimport_files(files_to_import);
+
+	uint32_t step = 1;
+	for (const String &file : paths) {
+		Ref<Resource> res = ResourceLoader::load(file);
+		ep.step(TTR("Attempting to re-save ") + file, step++);
+		if (res.is_valid()) {
+			// Ignore things that fail to load.
+			ResourceSaver::save(res);
+		}
+	}
+}
+
+void SurfaceUpgradeDialog::_show_popup() {
+	RS::get_singleton()->set_surface_upgrade_callback(nullptr);
+	if (EditorNode::immediate_confirmation_dialog(TTR("Attempting to use a mesh with an outdated mesh format. \nThe engine needs to update the format in order to use the mesh.\n\nPress 'Upgrade & Re-save' to have the engine scan the project folder and\nautomatically update and re-save all meshes and scenes.\nThis update may take a few minutes.\nUpgrading will make the meshes incompatible with previous versions of Godot.\n\nPress 'Upgrade Only' to continue opening the scene as normal. The engine will update\neach mesh in memory, but the update will not be saved.\nChoosing this option will lead to slower load times every time this project is loaded."), TTR("Upgrade & Re-save"), TTR("Upgrade Only"))) {
```

Comment:
Maybe move it to a variable so the conditional is more readable? 🙃

Reply:
```suggestion
	if (EditorNode::immediate_confirmation_dialog(TTR("Attempting to use a mesh with an outdated mesh format.\nThe engine needs to update the format in order to use the mesh.\n\nPress 'Upgrade & Re-save' to have the engine scan the project folder and\nautomatically update and re-save all meshes and scenes.\nThis update may take a few minutes.\nUpgrading will make the meshes incompatible with previous versions of Godot.\n\nPress 'Upgrade Only' to continue opening the scene as normal. The engine will update\neach mesh in memory, but the update will not be saved.\nChoosing this option will lead to slower load times every time this project is loaded."), TTR("Upgrade & Re-save"), TTR("Upgrade Only"))) {
```
Trailing space.

But the manual line breaks are also problematic and won't necessarily be translated properly, or scale well with different screen resolutions or editor scaling.

It would be better to use autowrapping, and limit the width of the dialog.

Reply:
It appears to be wrapped already, so all that is needed is to remove single newlines within paragraphs. Splitting paragraphs with doubled newlines is fine, and we do it in the project converter dialogs.

Reply:
Done!